firewood/src/hashednode.rs

```diff
@@ -310,43 +309,47 @@ impl HasUpdate for Vec<u8> {
     }
 }
 
-pub(super) enum ValueDigest<'a> {
-    /// A node's value.
-    Value(&'a [u8]),
-    /// The hash of a a node's value.
-    /// TODO danlaine: Use this variant when implement ProofNode
-    _Hash(Box<[u8]>),
+#[derive(Clone, Debug)]
+/// A ValueDigest is either a node's value or the hash of its value.
+pub enum ValueDigest<T> {
+    Value(T),
+    /// TODO this variant will be used when we deserialize a proof node
+    /// from a remote Firewood instance. The serialized proof node they
+    /// send us may the hash of the value, not the value itself.
+    _Hash(T),
 }
 
-trait Hashable {
+pub(crate) trait Hashable {
+    type T: AsRef<[u8]>;
+
     /// The key of the node where each byte is a nibble.
     fn key(&self) -> impl Iterator<Item = u8> + Clone;
     /// The node's value or hash.
-    fn value_digest(&self) -> Option<ValueDigest>;
+    fn value_digest(&self) -> Option<ValueDigest<Self::T>>;
     /// Each element is a child's index and hash.
     /// Yields 0 elements if the node is a leaf.
     fn children(&self) -> impl Iterator<Item = (usize, &TrieHash)> + Clone;
 }
 
 pub(super) trait Preimage {
     /// Returns the hash of this preimage.
-    fn to_hash(self) -> TrieHash;
+    fn to_hash(&self) -> TrieHash;
     /// Write this hash preimage to `buf`.
-    fn write(self, buf: &mut impl HasUpdate);
+    fn write(&self, buf: &mut impl HasUpdate);
 }
 
 // Implement Preimage for all types that implement Hashable
-impl<T> Preimage for T
+impl<T: Hashable> Preimage for T
 where
-    T: Hashable,
+    T::T: AsRef<[u8]>,
```

Comment:
Either use a `where` clause or bounds, not both

`impl<T: Hashable + AsRef<[u8]>>`

or
```
impl<T> Preimage for T
where
    T: Hashable,
    T: AsRef<[u8]>,
```

Reply:
change to use just the bounds